app/src/main/java/com/duckduckgo/app/httpsupgrade/api/HttpsUpgradeListDownloader.kt

```diff
@@ -38,21 +51,47 @@ class HttpsUpgradeListDownloader @Inject constructor(
             val call = service.https()
             val response = call.execute()
 
-            if (response.isCached && httpsUpgradeDao.count() != 0) {
+            if (response.isCached && dbWriteStatusStore.hasWriteCompleted() && httpsUpgradeDao.count() != 0) {
                 Timber.d("HTTPS data already cached and stored")
                 return@fromAction
             }
 
             if (response.isSuccessful) {
                 Timber.d("Updating HTTPS upgrade list from server")
-                val domains = response.body()!!.toTypedArray()
-                database.runInTransaction {
+
+                val domains = response.body() ?: throw IllegalStateException("Failed to obtain HTTPS upgrade list")
+
+                val startTime = System.currentTimeMillis()
+
+                modifyHttpsDatabase {
                     httpsUpgradeDao.deleteAll()
-                    httpsUpgradeDao.insertAll(*domains)
+                    Timber.v("Took ${System.currentTimeMillis() - startTime}ms to delete existing records")
+
+                    val chunks = domains.chunked(chunkSize)
+                    Timber.i("Received ${domains.size} HTTPS domains; chunking by $chunkSize into ${chunks.size} separate DB transactions")
+
+                    chunks.forEach {
+                        database.runInTransaction {
+                            httpsUpgradeDao.insertAll(it)
+                        }
+                    }
+
+                    Timber.i("Successfully wrote HTTPS data; took ${System.currentTimeMillis() - startTime}ms")
                 }
+
             } else {
                 throw IOException("Status: ${response.code()} - ${response.errorBody()?.string()}")
             }
         }
     }
+
+    private inline fun modifyHttpsDatabase(function: () -> Unit) {
+        dbWriteStatusStore.updateStatus(false)
+        function()
```

Comment:
This seems odd that it can run any function, right?

Reply:
I made the `modifyHttpsDatabase` function to hopefully draw enough attention to the fact that the action, whatever it may be, is immediately wrapped with calls to update the status store state.

It can run any function, you're right. But hopefully given it's existence in this class, it's clear enough that you'd want to wrap a bunch of code which is going to interact with the DB. This function could easily be removed and we could just leave the `dbWriteStatusStore.updateStatus()` calls top and bottom of the rest of the code, but I had hoped this would draw attention to them in the right way.

I can address if it's making it less clear?